Populate dashpages from configuration

drewnoakes opened the pull request:
Fixes #5289

This change causes the dashboard to look for a file named `dashpages.json` in the current directory. If found, it's parsed to provide the set of dashpages shown on the metrics page.


## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [x] No. Follow-up changes expected. (this is a PR to a feature branch)
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No (not yet, we will doc dashpages entirely before the feature branch merges)

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5302)

drewnoakes committed:
Populate dashpages from configuration

Fixes #5289

This change causes the dashboard to look for a file named `dashpages.json` in the current directory. If found, it's parsed to provide the set of dashpages shown on the metrics page.

radical commented on tests/Aspire.Dashboard.Tests/Model/DashpageJsonFilePersistenceTests.cs:
```diff
@@ -0,0 +1,105 @@
+            ]
+            """;
+
+        var dashpages = DashpageJsonFilePersistence.Deserialize(json);
```
Can the test use ` DashpageJsonFilePersistence.GetDashpagesAsync` instead since that is the public API?

drewnoakes replied:
That requires writing a file or mocking the file provider which I wanted to avoid.

radical replied:
Writing a json file to be loaded by the public api? It can have the same contents as in this existing test, right?

drewnoakes replied:
The main thing I want to test is the transformation between string and object model. The other API tests more layers than needed here.

radical commented on src/Aspire.Dashboard/Model/DashpageModels.cs:
```diff
@@ -0,0 +1,107 @@
+            {
+                using Stream stream = fileInfo.CreateReadStream();
+
+                using StreamReader reader = new(stream, Encoding.UTF8, leaveOpen: true);
```
Just curious - why is `leaveOpen: true` needed here?

drewnoakes replied:
Either way would be fine. I prefer to have the `using Stream` close the stream, and the reader not actually touch the stream when disposed, but closing the stream twice would also be safe.

drewnoakes commented:
/azp run

drewnoakes committed:
Test fixes

bunit doesn't support when `OnInitializedAsync` actually yields the thread, so provide bunit tests with a stub implementation of `IDashpagePersistence.GetDashpagesAsync` that returns an empty set synchronously. In future we may extend this type to allow test setup with dashpage definitions.

adamint commented on src/Aspire.Dashboard/DashboardWebApplication.cs:
```diff
@@ -219,6 +220,9 @@ public DashboardWebApplication(
 
         builder.Services.AddScoped<DimensionManager>();
 
+        builder.Services.AddSingleton<IFileProvider>(new PhysicalFileProvider(Directory.GetCurrentDirectory()));
```
why is this necessary?

drewnoakes replied:
We're the first consumer of the `IFileProvider` abstraction. It allows the other service `DashpageJsonFilePersistence` access to the file system.

adamint commented on src/Aspire.Dashboard/dashpages.json:
```diff
@@ -0,0 +1,38 @@
+[
+  /* Comments are ignored */
```
What do you think about documenting this schema elsewhere, and removing the comments? It is a good practice to avoid commenting JSON, even though our deserializer currently handles it.

drewnoakes replied:
Comments were just because I copied this from a unit test where I wanted to validate that comments are tolerated.

Adding a schema makes sense.

adamint replied:
I do not think we should officially support JSON with comments. It is not valid JSON.

drewnoakes replied:
It's very common to want to document your configuration. It's not very friendly to disallow that when it's so easy to permit it. I see no downside.

drewnoakes replied:
For example, the .NET Project System ignores comments in `launchSettings.json` files.

drewnoakes commented:
`Aspire.EndToEnd.Tests.IntegrationServicesTests.Aspire.EndToEnd.Tests.IntegrationServicesTests.VerifyCosmosComponentWorks(resourceName: efcosmos)` timed out after 10 minutes.

drewnoakes commented:
/azp run
